uxarray/_mpas.py

```diff
@@ -0,0 +1,266 @@
+            "units": "degrees_north",
+        })
+
+    # vertex indices that surround each primal-mesh cell
+    verticesOnCell = in_ds['verticesOnCell'].values
+
+    # max number of edges & vertices that make up each primal-mesh cell
+    nEdgesOnCell = in_ds['nEdgesOnCell'].values
+
+    # correct to be zero-indexed and to use fill values
+    verticesOnCell_fill = _add_fill_values(verticesOnCell, nEdgesOnCell)
+
+    out_ds["Mesh2_face_nodes"] = xr.DataArray(
+        data=verticesOnCell_fill,
+        dims=["nMesh2_face", "nMaxMesh2_face_nodes"],
+        attrs={
+            "cf_role": "face_node_connectivity",
+            "_FillValue": fill_val,
+            "start_index": int_dtype(0)
+        })
+
+    # vertex indices that saddle a given edge
+    verticesOnEdge = in_ds['verticesOnEdge'].values
+
+    # convert to zero-indexed
+    if verticesOnEdge.min() == 1:
+        verticesOnEdge -= 1
+
+    out_ds["Mesh2_edge_nodes"] = xr.DataArray(
+        data=verticesOnEdge,
+        dims=["nMesh2_edge", "Two"],
+        attrs={
+            "cf_role": "edge_node_connectivity",
+            "start_index": int_dtype(0)
+        })
+
+
+def _dual_to_ugrid(in_ds, out_ds):
+    """If input dataset is an unstructured MPAS file, function will encode the
+    MPAS Dual-Mesh in the UGRID conventions.
+
+    Parameters
+    ----------
+    in_ds : xarray.Dataset
+        Input MPAS dataset
+    out_ds : xarray.Dataset
+        Output dataset where the MPAS Dual-Mesh is encoded in the UGRID
+        conventions
+    """
+
+    # corners of dual-mesh cells (in degrees)
+    lonCell = np.rad2deg(in_ds['lonCell'].values)
+    latCell = np.rad2deg(in_ds['latCell'].values)
+
+    out_ds['Mesh2_node_x'] = xr.DataArray(
+        lonCell,
+        dims=["nMesh2_node"],
+        attrs={
+            "standard_name": "longitude",
+            "long_name": "longitude of mesh nodes",
+            "units": "degrees_east",
+        })
+
+    out_ds['Mesh2_node_y'] = xr.DataArray(
+        latCell,
+        dims=["nMesh2_node"],
+        attrs={
+            "standard_name": "latitude",
+            "long_name": "latitude of mesh nodes",
+            "units": "degrees_north",
+        })
+
+    # centers of dual-mesh cells (in degrees)
+    lonVertex = np.rad2deg(in_ds['lonVertex'].values)
+    latVertex = np.rad2deg(in_ds['latVertex'].values)
+
+    out_ds['Mesh2_face_x'] = xr.DataArray(
+        lonVertex,
+        dims=["nMesh2_face"],
+        attrs={
+            "standard_name": "longitude",
+            "long_name": "longitude of center nodes",
+            "units": "degrees_east",
+        })
+
+    out_ds['Mesh2_face_y'] = xr.DataArray(
+        latVertex,
+        dims=["nMesh2_face"],
+        attrs={
+            "standard_name": "latitude",
+            "long_name": "latitude of center nodes",
+            "units": "degrees_north",
+        })
+
+    # vertex indices that surround each dual-mesh cell
+    cellsOnVertex = in_ds['cellsOnVertex'].values
```

Comment:
For the dual mesh, we need to talk about how to handle vertices with fewer than 3 neighboring cells.  These are indicated by zeros in `cellsOnVertex`.  The occur in planar or spherical meshes where some cells have been removed (e.g. land cells removed from the ocean).  For visualization purposes, a vertex with one or more invalid neighboring cells is degenerate.  But for mathematical purposes, they remain useful.  I don't know how uxarray plans to handle such situations.

In any case, some code needs to be added to convert zero values to mask values here.

Reply:
This is a great point. I think this would be a good candidate for a helper function.

Reply:
I think your call to `_replace_zeros()` has taken care of this nicely.